runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -6546,14 +6547,37 @@ done:;
 				restoreGenericSpecialStackFrame(REGISTER_ARGS);
 				if (immediateAsyncPending()) {
 					rc = GOTO_ASYNC_CHECK;
+					goto exit;
 				} else if (VM_VMHelpers::exceptionPending(_currentThread)) {
 					rc = GOTO_THROW_CURRENT_EXCEPTION;
+					goto exit;
 				} else {
 					profileCallingMethod(REGISTER_ARGS);
 					_sendMethod = method;
 				}
 			}
+
+			/* Check the receiver class is the subtype of the current class [= interface] */
+			J9Class *resolvedClass = J9_CLASS_FROM_METHOD(_sendMethod);
+			J9ROMMethod *romMethod = J9_ROM_METHOD_FROM_RAM_METHOD(_sendMethod);
+			/* Skip the check for final methods in java.lang.Object (resolvedClassDepth == 0) */
+			if ((0 != J9CLASS_DEPTH(resolvedClass)) || J9_ARE_NO_BITS_SET(romMethod->modifiers, J9AccFinal)) {
+				J9Class *receiverClass = J9OBJECT_CLAZZ(_currentThread, receiver);
+				J9Class *currentClass = J9_CLASS_FROM_CP(ramConstantPool);
+				/* hostClass is exclusively defined only in Unsafe.defineAnonymousClass.
+				 * For all other cases, clazz->hostClass points to itself (clazz).
+				 */
+				currentClass = currentClass->hostClass;
+				if (!instanceOfOrCheckCast(receiverClass, currentClass)) {
+					buildMethodFrame(REGISTER_ARGS, _sendMethod, 0);
+					updateVMStruct(REGISTER_ARGS);
+					setIllegalAccessErrorReceiverNotSameOrSubtypeOfCurrentClass(_currentThread, receiverClass, currentClass);
+					VMStructHasBeenUpdated(REGISTER_ARGS);
+					rc = GOTO_THROW_CURRENT_EXCEPTION;
+				}
+			}
 		}
+exit:
```

Comment:
To keep things similar to the rest of the file, please use "done" as the terminal label.

Reply:
should I also change the label for `invokespecial` as well then (these are the same)?

Reply:
Please - this problem will go away when you merge the two invokespecials together.